File: CPP/3-Two Pointers/4-container_max_area.cpp
#include <iostream>
#include <vector>

using namespace std;

int maxArea(vector<int>& height) {
    int max_area = 0;
    int len = height.size();
    int left = 0;
    int right = len - 1;

    while (left < right) {
        int area = min(height[left], height[right]) * (right - left);
        if (area > max_area) {
            max_area = area;
        }
        
        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }

    return max_area;
}

int main() {
    vector<int> height = {1,8,6,2,5,4,8,3,7};

    int area = maxArea(height);
    cout << area;

    return 0;
}

File: PYTHON/3-TwoPointers/4-container_max_area.py
from typing import List

def maxArea(heights: List[int]) -> int:
    left, right = 0, len(heights) - 1
    max_area = 0

    while left < right:
        area = min(heights[left], heights[right]) * (right - left)
        if area > max_area:
            max_area = area
        
        if heights[left] < heights[right]:
            left += 1
        else:
            right -= 1
    
    return max_area


if __name__ == "__main__":
    heights = [1, 7, 2, 5, 4, 7, 3, 6]
    print(maxArea(heights))
        
